### PR TITLE
Use filter for azuread_service_principal and azuread_application data sources 

### LINKED ISSUE #1844: azurerm_azuread_service_principal does not always find a Service Principal
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.8
+ provider.azurerm v1.13.0

### Affected Resource(s)

Data Source: azurerm_azuread_service_principal

### Terraform Configuration Files

```
variable "application_name" {
  type = "string"
}

resource "azurerm_azuread_application" "test" {
  name = "${var.application_name}"
}

resource "azurerm_azuread_service_principal" "test" {
  application_id = "${azurerm_azuread_application.test.application_id}"
}

data "azurerm_azuread_service_principal" "test" {
  application_id = "${azurerm_azuread_application.test.application_id}"
}

output "spn_object_id" {
  value = "${data.azurerm_azuread_service_principal.test.id}"
}
```

### Debug Output

[gist](https://gist.github.com/steve-hawkins/ee2c3b25a96d328631ac4fd338113fac)

### Panic Output

N/A

### Expected Behavior

The Application Registration Service Principal Object ID should be returned

### Actual Behavior

A Service Principal for Application ID was not found

### Steps to Reproduce

This part is hard to document as I have only noticed it on three out of five current Application Registrations.  I will add more here if I can find an issue with the actual Application Registrations, but as I mention in the section below using the Azure CLI I can find the Service Principal Object ID for all the five Application Registrations I have created.

1. `terraform apply`

### Important Factoids

`az ad sp show --id '[REDACTED]'`

```
{
  "accountEnabled": true,
  "addIns": [],
  "alternativeNames": [],
  "appDisplayName": "test_application",
  "appId": "[REDACTED]",
  "appOwnerTenantId": "[REDACTED]",
  "appRoleAssignmentRequired": false,
  "appRoles": [],
  "deletionTimestamp": null,
  "displayName": "test_application",
  "errorUrl": null,
  "homepage": "http://test_application",
  "keyCredentials": [],
  "logoutUrl": null,
  "oauth2Permissions": [
    {
      "adminConsentDescription": "Allow the application to access test_application on behalf of the signed-in user.",
      "adminConsentDisplayName": "Access test_application",
      "id": "[REDACTED]",
      "isEnabled": true,
      "type": "User",
      "userConsentDescription": "Allow the application to access test_application on your behalf.",
      "userConsentDisplayName": "Access test_application",
      "value": "user_impersonation"
    }
  ],
  **"objectId": "[REDACTED]",
  "objectType": "ServicePrincipal",**
  "odata.metadata": "https://graph.windows.net/[REDACTED]/$metadata#directoryObjects/Microsoft.DirectoryServices.ServicePrincipal/@Element",
  "odata.type": "Microsoft.DirectoryServices.ServicePrincipal",
  "passwordCredentials": [],
  "preferredTokenSigningKeyThumbprint": null,
  "publisherName": "Steve Limited",
  "replyUrls": [],
  "samlMetadataUrl": null,
  "servicePrincipalNames": [
    "http://test_application",
    "[REDACTED]"
  ],
  "servicePrincipalType": "Application",
  "signInAudience": "AzureADMyOrg",
  "tags": [],
  "tokenEncryptionKeyId": null
}
```

### References

N/A

### PR DESCRIPTION
**Pull request by tiwood**
This fixes #1844

This adds a filter to the ListAll() method to limit the results that are returned by the API to only include objects that match the data source query.

### CODE REVIEW

File: azurerm/data_source_azuread_application.go [hunk 1]
```diff
@@ -85,13 +87,18 @@ func dataSourceArmAzureADApplicationRead(d *schema.ResourceData, meta interface{
 
 		application = resp
 	} else {
-		resp, err := client.ListComplete(ctx, "")
+
+		// use the name to find the Azure AD application
+		name := d.Get("name").(string)
+		filter := "displayName eq '" + name + "'"
```
Comment (tombuildsstuff) [inline azurerm/data_source_azuread_application.go]: this'd be better as `filter := fmt.Sprintf("displayName eq '%q'", name)

File: azurerm/data_source_azuread_service_principal.go [hunk 2]
```diff
@@ -55,47 +57,61 @@ func dataSourceArmActiveDirectoryServicePrincipalRead(d *schema.ResourceData, me
...
 
 		servicePrincipal = &app
-	} else {
-		apps, err := client.ListComplete(ctx, "")
+
+	} else if _, ok := d.GetOk("display_name"); ok {
+
+		// use the display_name to find the Azure AD service principal
+		displayName := d.Get("display_name").(string)
+		filter := "displayName eq '" + displayName + "'"
```
Comment (tombuildsstuff) [inline azurerm/data_source_azuread_service_principal.go]: (as above) could we make this `filter := fmt.Sprintf("displayName eq '%q'", displayName)`

File: azurerm/data_source_azuread_service_principal.go [hunk 3]
```diff
@@ -55,47 +57,61 @@ func dataSourceArmActiveDirectoryServicePrincipalRead(d *schema.ResourceData, me
...
+
+		if servicePrincipal == nil {
+			return fmt.Errorf("A Service Principal with the Display Name %q was not found", displayName)
+		}
+
+	} else {
+
+		// use the application_id to find the Azure AD service principal
+		applicationId := d.Get("application_id").(string)
+		filter := "appId eq '" + applicationId + "'"
```
Comment (tombuildsstuff) [inline azurerm/data_source_azuread_service_principal.go]: (as above) could we use `fmt.Sprintf("appId eq '%q'", applicationId)` here?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @tiwood 

Thanks for this PR - I've taken a look through and this mostly LGTM, if we can fix up the (minor) comments we should be able to run the tests and get this merged :)

Thanks!

File: azurerm/data_source_azuread_application.go (hunk 1 above)
Comment (tiwood) [reply azurerm/data_source_azuread_application.go]: Makes sense! Added as requested.

File: azurerm/data_source_azuread_service_principal.go (hunk 2 above)
Comment (tiwood) [reply azurerm/data_source_azuread_service_principal.go]: Makes sense! Added as requested.

File: azurerm/data_source_azuread_service_principal.go (hunk 3 above)
Comment (tiwood) [reply azurerm/data_source_azuread_service_principal.go]: Makes sense! Added as requested.

**Review by tiwood — COMMENTED** [review]
Made changes as requested. Thank you for the review.

**Review by tombuildsstuff — APPROVED** [review]
hey @tiwood 

Thanks for pushing those changes - this now LGTM 👍 

Thanks!

**Comment by tombuildsstuff** [comment]
Tests pass:

screenshot 2018-09-07 at 20 30 38

**Pull request merged**